### PR TITLE
fix(#1036): verify a resumed download was honoured (206 + Content-Range) before appending

### LINKED ISSUE #1036: ModelDownloader: resume appends the full body after a stale .part when the server ignores Range (corrupt model that may still LOAD)
Found while verifying #1029, reproduced deterministically with a control.

## The bug

`startDownload` resumes from any existing `<dest>.part` by sending `Range: bytes=<size>-`. But **nothing checks that the response honoured it.** The only `Content-Range` handling is progress arithmetic in `onDownloadProgress` (adds `m_resumeOffset` to the totals); `HttpStatusCodeAttribute` (206 vs 200) and the actual `Content-Range` header are never inspected.

So any response that ignores `Range` — `file://` **always** does; any CDN/proxy that strips the header will — returns the **full body with a 200**, and `onReadyRead` appends it after the stale prefix.

## Reproduction (lama.onnx, 208,044,816 bytes)

| stale `.part` | result | |
|---|---|---|
| 29 bytes seeded | **208,044,845 = 29 + 208,044,816**, begins with the seed text | corrupt |
| none (control) | 208,044,816, SHA-256 identical to the real file | clean |

## Why this is worse than a load error

The two corrupted copies behaved **differently**:
- a 30-byte stale prefix → ONNX Runtime: `Protobuf parsing failed` (detected at load)
- a 29-byte prefix → **loaded fine, ran, and produced byte-identical output** to the clean model (ORT parsed the garbage as an unknown protobuf field)

Whether corruption is even *detected* at load depends on what the stray bytes happen to parse as. A successful load proves nothing about integrity — which is the CWE-494 point #1029 makes, and why its SHA-256 check is the actual safety net: with a digest, both copies are deleted instead of cached.

## Fix

In `onDownloadFinished` (or on first `readyRead`), when `m_resumeOffset > 0`:
- read `HttpStatusCodeAttribute`; if it is **200** (not 206) or `Content-Range` is absent/doesn't start at `m_resumeOffset`, the server sent the whole body — **truncate the `.part` to 0 and restart the write** (or discard and re-download), never append.
- for `file://` there is no status; treat resume as unsupported and always start from 0.

Then the stale-`.part` case becomes self-healing instead of self-poisoning. Add a `FakeNetworkReply` test that sets the status attribute to 200 with a non-zero `m_resumeOffset` and asserts the result equals the payload alone.

### PR DESCRIPTION
**Pull request by fernandotonon**
Closes #1036. Found and reproduced while verifying #1029.

## The bug

A resume sends `Range: bytes=N-`, but **nothing checked that the server honoured it**. The only `Content-Range` handling was progress arithmetic. A server that ignores `Range` — `file://` always does; any proxy/CDN that strips the header will — answers **200 with the whole body**, and `onReadyRead` appended it after the stale `.part`.

Reproduced deterministically, with a control (`lama.onnx`, 208,044,816 bytes):

| stale `.part` | result | |
|---|---|---|
| 29 bytes seeded | **208,044,845 = 29 + full body**, begins with the seed text | corrupt |
| none (control) | 208,044,816, SHA-256 identical to the real file | clean |

**The alarming part:** the 29-byte-corrupted model **loaded, ran, and produced byte-identical output** to the clean one — ORT parsed the garbage as an unknown protobuf field. A 30-byte prefix, by contrast, failed with `Protobuf parsing failed`. Whether corruption is even *detected* at load is arbitrary. A successful load proves nothing about integrity, which is why #1029's digest check exists — but the downloader should not be manufacturing corrupt files in the first place.

## Fix

On the **first `readyRead`** of a resumed request (a flag armed at both sites that send a `Range` header), require status **206** *and* a `Content-Range` whose first byte equals our resume offset. Anything else is treated as the full body: reopen the `.part` with `Truncate`, zero `m_resumeOffset`/`m_bytesReceived` (otherwise progress adds a phantom offset), and continue from byte 0.

Two deliberate choices:
- **No `downloadError`.** A 200 is recoverable — the download still completes correctly, just from the start. Failing it would turn a self-healing case into a user-visible one.
- **A 206 whose window starts elsewhere is also rejected.** The server chose a different range; appending it would land bytes in the wrong place, which is the same corruption by another route.

Checked once, not per chunk — status and headers are available at the first chunk and re-checking is waste.

## Tests

- 200 → restarts from zero: payload alone on disk, offset 0, no error
- honoured 206 → appends and **keeps** its offset (the existing behaviour, pinned)
- 206 with a wrong start → restarts from zero
- a fresh (non-resume) download never runs the check

`FakeNetworkReply` now defaults to a plain 200 — exactly what an ignoring server returns — and gains `withPartialContent(first,last,total)` to model an honoured resume. `setAttribute` is protected on `QNetworkReply`, so the subclass can inject the status.

**Mutation-verified with correct selectivity:** disabling the check fails the 200 test; dropping *only* the offset comparison fails *only* the wrong-start test and leaves the honoured and 200 tests green. The tests discriminate the two guards rather than tripping on any change — an over-coupled test would have failed the wrong case under the second mutant, and I checked for that explicitly. Suite 35/35 under `--gtest_shuffle` ×3.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved resumed downloads by verifying that servers honor requested byte ranges.
  * Automatically restarts downloads from the beginning when range requests are ignored or return the complete file.
  * Stops and removes incomplete files when servers return an unrelated partial range.
  * Prevents incomplete downloads from being marked as complete when their size does not match the expected total.
  * Preserves valid partial files for later resumption when a transfer cannot be completed.
  * Ensures range-related failures produce a single download error.
  * Supports case-insensitive range units for more reliable resume behavior during paused transfers.

### CODE REVIEW

**Comment by fernandotonon** [comment]
Valid, and it led somewhere — addressed in a9ad66a2.

**The finding itself:** RFC 9110 §14.1 makes the range unit case-insensitive, so `Content-Range: Bytes 9-12/13` is a perfectly honoured resume. My `startsWith("bytes ")` misread it as "server ignored Range" and fell into the truncate path *with a partial body* — producing an incomplete file that a no-digest caller would rename and cache. That is strictly worse than the bug this PR set out to fix. The unit is now compared with `Qt::CaseInsensitive`, and `Content-Range` is parsed fully (first/last/total).

**What following the consequence exposed in my own code:** the fallback "truncate and take the body from 0" is only safe when the body is the *full* resource. A 206 whose window is neither ours nor `0..total-1` is a genuinely partial body we didn't ask for — and my previous handling wrote it from 0, which would have produced an incomplete file by the same route your scenario describes. That case now **aborts**: `downloadError`, `.part` removed so the next attempt starts clean, nothing renamed. Only a 206 covering exactly `0..total-1` (a full body wearing a partial status) is truncated-and-taken like a 200.

**Tests:** `Bytes 9-12/13` appends and keeps its offset; a foreign window `3-5/10` emits an error and leaves no `.part`; the whole-resource `0-2/3` test is kept but re-titled to state its real intent (it was previously passing for a misleading reason).

**Mutation, both with exact selectivity:** a case-sensitive compare fails *only* the Bytes test; dropping the whole-resource detection fails *only* the whole-resource test and leaves the foreign-window abort green. 37/37 under `--gtest_shuffle` ×3.

**Comment by fernandotonon** [comment]
CI was fully green; the merge gate held on **SonarCloud's maintainability rating** — `onReadyRead` had grown to cognitive complexity 36 (limit 25) with the `Content-Range` parsing nested inline, plus `qsizetype→int` narrowing and multi-declaration smells. Structure only, no functional finding.

Addressed in 813281ef by extracting a pure `parseContentRange` (now **unit-tested directly** — case-insensitive unit, `*` total, whitespace, nine malformed shapes — rather than only through the resume behaviour tests) and a `verifyResumeResponse()` method; `onReadyRead` is four lines.

Behaviour unchanged by construction and by evidence: 39/39 under `--gtest_shuffle` ×3, and the two mutants from the earlier round (case-sensitive unit; whole-resource detection dropped) are still caught with the same selectivity — the first now fails both the behavioural test *and* the new parser test.

File: src/ModelDownloader.cpp [hunk 1]
```diff
@@ -313,10 +315,93 @@ void ModelDownloader::cancelDownload()
...
+                    first = range.left(dash).trimmed().toLongLong(&ok1);
+                    last  = range.mid(dash + 1, slash - dash - 1).trimmed().toLongLong(&ok2);
+                    total = range.mid(slash + 1).trimmed().toLongLong(&ok3);   // "*" fails ok3 => -1
+                }
+                if (!ok1) first = -1;
+                if (!ok2) last = -1;
+                if (!ok3) total = -1;
+            }
+        }
+        const bool honoured = status == 206 && first == m_resumeOffset;
```
[the remark this answers is not in the document]
Comment (fernandotonon) [reply src/ModelDownloader.cpp]: Re-checked against the code, not the annotation: this was NOT addressed in 813281e — the honoured check was still `first == m_resumeOffset` alone. Fixed in 693264d2: honoured now also requires `last == total-1`; `bytes 9-10/13` goes through the discard path. Test: `ResumeAgainst206ShortWindowIsRejectedNotAppended`.

File: src/ModelDownloader.cpp [hunk 2]
```diff
@@ -313,10 +315,93 @@ void ModelDownloader::cancelDownload()
...
+    // and appending that after the stale .part produced a corrupt model —
+    // reproduced: a 29-byte stale prefix yielded a 208,044,845-byte file that
+    // still LOADED and ran (ORT parsed the garbage as an unknown protobuf
+    // field). A load success proves nothing; only the response can tell us
+    // which bytes these are. So on the first bytes of a resumed request we
+    // require 206 + a Content-Range that starts exactly at our offset;
+    // anything else means "this is the full body" — truncate and restart
+    // from byte 0, never append. Checked ONCE, on first readyRead, because the
+    // status/headers are available then and re-checking per chunk is waste.
+    if (m_resumeUnverified) {
```
[the remark this answers is not in the document]
Comment (fernandotonon) [reply src/ModelDownloader.cpp]: Valid — fixed in 693264d2. `onDownloadFinished` refuses to promote while `m_resumeUnverified` is still set (the resume reply finished without ever delivering data, so the check in `onReadyRead` never ran); the stale `.part` is kept for the next resume and one `downloadError` is emitted. Test: `FinishedWithoutBodyWhileResumeUnverifiedDoesNotRename`.

File: src/ModelDownloader.cpp [hunk 3]
```diff
@@ -311,12 +313,107 @@ void ModelDownloader::cancelDownload()
...
+    // "recover".
+    if (const bool fullBodyDisguised = parsed && first == 0 && total > 0 && last == total - 1;
+        status == 206 && !fullBodyDisguised) {
+        qCritical() << "ModelDownloader: unusable 206 for" << m_currentModelName
+                    << "— asked from byte" << m_resumeOffset << "got Content-Range first="
+                    << first << "last=" << last << "total=" << total << "— aborting";
+        m_outputFile->close();
+        QFile::remove(m_tempFilePath);   // next attempt must start clean, not resume this
+        m_currentReply->abort();
+        emit downloadError(m_currentModelName,
```
[the remark this answers is not in the document]
Comment (fernandotonon) [reply src/ModelDownloader.cpp]: Valid — fixed in 693264d2. All 'unusable partial' exits go through one `discardPartialAndFail`: the reply is detached and aborted under `m_abortingInternally`, so the synchronously-delivered `errorOccurred`/`finished` step aside and exactly one `downloadError` is emitted. The test's `FakeNetworkReply` now signals from `abort()` and is wired like production — the first version of that test was vacuous (unwired) and a mutant with the guard removed passed it; the fixed test fails the mutant with `errors == 2`.

File: src/ModelDownloader.cpp [hunk 4]
```diff
@@ -311,12 +313,107 @@ void ModelDownloader::cancelDownload()
...
+    m_resumeUnverified = false;
+    const int status = m_currentReply->attribute(
+        QNetworkRequest::HttpStatusCodeAttribute).toInt();
+    qint64 first = -1;
+    qint64 last = -1;
+    qint64 total = -1;
+    const bool parsed = status == 206
+        && parseContentRange(m_currentReply->rawHeader("Content-Range"), first, last, total);
+    if (parsed && first == m_resumeOffset)
+        return true;   // honoured: keep appending
```
[the remark this answers is not in the document]
Comment (fernandotonon) [reply src/ModelDownloader.cpp]: Valid — fixed in 693264d2. The response's committed size is recorded (`m_expectedTotalBytes`: the 206 total, else a full body's Content-Length) and the finished `.part` size must match it before the rename; a short body leaves the `.part` in place for the next resume and emits one error. One deliberate deviation: an UNKNOWN size with no digest is accepted with a `qWarning` rather than refused — that would break chunked-transfer servers, and HF/GitHub plus QNAM's `file://` backend always send Content-Length, so real downloads take the strict path. Tests: `HonouredResumeWithShortBodyIsNotRenamed`, `FreshDownloadShorterThanContentLengthIsNotRenamed`, `UnknownLengthWithoutDigestIsAcceptedWithWarning`.
